File: backend/app/services/hasher.py
from __future__ import annotations

import hashlib
from datetime import date


def compute_transaction_hash(
    txn_date: date,
    amount_cents: int,
    raw_description: str,
    account_id: int,
) -> str:
    """SHA-256 hash for transaction deduplication.

    Uses raw_description (not cleaned) to ensure exact-match semantics.
    Uses account_id to allow the same transaction description+amount on
    the same date in different accounts.
    """
    payload = f"{txn_date.isoformat()}|{amount_cents}|{raw_description}|{account_id}"
    return hashlib.sha256(payload.encode("utf-8")).hexdigest()


def compute_file_hash(content: bytes) -> str:
    """SHA-256 hash of file content for whole-file deduplication."""
    return hashlib.sha256(content).hexdigest()

File: frontend/src/lib/local-engine/hasher.ts
export async function computeTransactionHash(
  date: string,
  amountCents: number,
  rawDescription: string,
  accountId: number
): Promise<string> {
  const payload = `${date}|${amountCents}|${rawDescription}|${accountId}`;
  const data = new TextEncoder().encode(payload);
  const hashBuffer = await crypto.subtle.digest("SHA-256", data);
  const hashArray = Array.from(new Uint8Array(hashBuffer));
  return hashArray.map((b) => b.toString(16).padStart(2, "0")).join("");
}

export async function computeFileHash(content: ArrayBuffer): Promise<string> {
  const hashBuffer = await crypto.subtle.digest("SHA-256", content);
  const hashArray = Array.from(new Uint8Array(hashBuffer));
  return hashArray.map((b) => b.toString(16).padStart(2, "0")).join("");
}
